### ISSUE TITLE
[8.0.x] war extender scans too many transitive dependencies

### ISSUE THREAD

**Issue by jutzig** [ops4j/org.ops4j.pax.web#1852]
After updating to pax-web 8, we noticed that the war deployments take way longer than before.
It seems the problem is caused by the way BundleWebApplicationClassSpace analyzes all wires of a bundle, but then continues the same process transitively for wires of wires.

We are using a large karaf container with a few thousand bundles, and the transient scanning seems end up analyzing almost every deployed bundle several times over. 
Deployment of all wars takes about 260 seconds with pax-web 8. After removing this line below, to prevent the transitive analysis, the deployment time was reduced to  44 seconds.
https://github.com/ops4j/org.ops4j.pax.web/blob/main/pax-web-extender-war/src/main/java/org/ops4j/pax/web/extender/war/internal/model/BundleWebApplicationClassSpace.java#L386

As discussed in slack, ideally it should ideally only take direct dependencies into consideration, not further down.
I'll create a pull request soon.

**Comment by jutzig** [comment]
So there were two places that traversed the complete dependency tree: BundleWebApplicationClassSpace and ClasspathUtil#getBundlesInClassSpace. In my opinion, that ClasspathUtil traverses all transitive dependencies does not match the javadoc of the method so it might make sense to remove the recursion there. But since it's public API and may have side effects, I wanted your take on that first. For now, I've created a non-recursive private copy of the method.

I did some performance tests with different fix permutations:

```
release:
	container boot time 12:40
	pax-web deploy time 05:40
	
fix in ClasspathUtil
	container boot time 05:40
	pax-web deploy time 01:00
	
fix in BundleWebApplicationClassSpace
	container boot time 06:30
	pax-web deploy time 01:25
	
both fixes:	
        container boot time 04:40
	pax-web deploy time 00:40
```

**Comment by grgrzybek** [comment]
Thanks for the tests - give me few days and I still think we can have 8.0.20 next week.

**Comment by jutzig** [comment]
Sure thing, thank you very much. I'll send the commit as a pull request once you let me know how you would like the ClasspathUtil handled.

**Comment by grgrzybek** [comment]
There were some test failures related to https://github.com/ops4j/org.ops4j.pax.web/tree/main/samples/samples-war/war-most-complex
I've created this complex scenario in the beginning of Pax Web 8 refactoring. I needed a WAB that has bundles wired, fragments and these fragments adding new requirements (treated as 1st level), but also requirements of requirements.

With this improvement, I had to stop treating `container-bundle-3` as reachable and adjust the tests.

However there's one additional JSF/Facelets/Primefaces integration test failing. And the exception is:
```
Caused by: java.lang.IllegalStateException: No Factories configured for this Application. This happens if the faces-initialization does not work at all - make sure that you properly include all configuration settings necessary for a basic faces application and that all the necessary libs are included. Also check the logging output of your web application and your container for any exceptions!
If you did that and find nothing, the mistake might be due to the fact that you use some special web-containers which do not support registering context-listeners via TLD files and a context listener is not setup in your web.xml.
A typical config looks like this;
<listener>
  <listener-class>org.apache.myfaces.webapp.StartupServletContextListener</listener-class>
</listener>

	at javax.faces.FactoryFinder._getFactory(FactoryFinder.java:310) ~[?:?]
	at javax.faces.FactoryFinder.getFactory(FactoryFinder.java:230) ~[?:?]
	at javax.faces.webapp.FacesServlet.init(FacesServlet.java:121) ~[?:?]
	at org.ops4j.pax.web.service.spi.servlet.OsgiInitializedServlet.init(OsgiInitializedServlet.java:68) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHolder$Wrapper.init(ServletHolder.java:1345) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHolder.initServlet(ServletHolder.java:632) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHolder.getServlet(ServletHolder.java:486) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHolder.prepare(ServletHolder.java:759) ~[?:?]
	at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHolder.prepare(PaxWebServletHolder.java:295) ~[?:?]
	at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler.doHandle(PaxWebServletHandler.java:316) ~[?:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) ~[?:?]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600) ~[?:?]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[9.4.51.v20230217:9.4.51.v20230217]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235) ~[?:?]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624) ~[?:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[?:?]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440) ~[?:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[?:?]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505) ~[?:?]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594) ~[?:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[?:?]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355) ~[?:?]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[?:?]
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191) ~[?:?]
	at org.ops4j.pax.web.service.jetty.internal.PrioritizedHandlerCollection.handle(PrioritizedHandlerCollection.java:96) ~[?:?]
	... 12 more
```

The problem is that there's no explicit wiring from `org.apache.myfaces.trinidad:trinidad-demo` to `org.apache.myfaces.core:myfaces-impl`, which contains `org.apache.myfaces.ee.MyFacesContainerInitializer` SCI...

**Comment by grgrzybek** [comment]
@jutzig thanks for the report.
Initially I took your https://github.com/jutzig/org.ops4j.pax.web/commit/a66cd3f64c6ee7cc9a150c75087fa70cef8cb7ca and verified it's ok. But then when working on new test failures, I had to rewrite the change a bit.

Would you be so kind to check current version (8.0.20-SNAPSHOT - you'd have to build it first...) and verify the results?

After some thinking I decided to turn off the recursion entirely from the point of view of WAB processing.

**Comment by jutzig** [comment]
Yeah, sorry about that, I had to locally tweak the pom a bit to make this work in my local karaf installation and therefore the full test suite couldn't run, so I missed this issue.
I see what you did there with the 'reached', that makes the code change much smaller, great!
Thanks a lot for the commit, looks perfect. I'll pull this, test it with our deployment and will report back as soon as I have result.

**Comment by grgrzybek** [comment]
Cool - please report the times (start + deploy).

**Comment by jutzig** [comment]
I ran some more tests with your commit and the performance is essentially identical to my initial commit:

```
new fix:	
    container boot time 04:30
    pax-web deploy time 00:40
```
Everything works exactly as expected, so thanks a lot for the quick fix!

**Comment by grgrzybek** [comment]
40s deploy is still much (IMO) - you can search (at `TRACE` level) for log lines with `Scanning ordered fragment` - how many jars are scanned?

You can then skip the JARs (by file name) or bundles (by symbolic name) you don't want to be scanned.

**Comment by jutzig** [comment]
> how many jars are scanned?
319. But it is a very large software so that number isn't unexpected.

**Comment by grgrzybek** [comment]
1. If some of the 319 don't have to be scanned, you can skip them using `org.ops4j.pax.web.extender.war.jarsToSkip` property
2. If all have to be scanned, then:
   1. respect for maintaining such big project (for me personally, that's the best part of my work)
   2. I'm glad Pax Web can handle it ;)

**Comment by jutzig** [comment]
Thanks for the hint. I looked more into it. Especially searching for lines like this helps to identify which pieces take the longest:
Finished metadata and fragment processing for io.hawt.hawtio-osgi [1016] in 2896ms
I used that to narrow down some heavy hitters that I could exclude (especially hibernate). That is my current exclude list:

`org.ops4j.pax.web.extender.war.jarsToSkip=default,org.hibernate.*,com.google.guava.*,com.fasterxml.*,groovy-*.jar`

That safes a few seconds, but it's not much and at the end of the day, it's just a lot of content. Especially the apereo CAS overlay war is huge and it takes some time to scan. I'll keep adding a few more jars/bundles to the exclude list, but for now I'm just happy that your fix brought a massive boost :-)
